### PR TITLE
Proof of concept: Everything starting with GameProblem until HumanKeyboardPlayer + game loop

### PR DESCRIPTION
**Pull request by TermanEmil**
* Define Abstract `GameProblem`
* Define `ChineseCheckersGameProblem`
* Define `Action` as a series of steps
* Define State as the current board + the id of the player's turn
* Define abstract Player + implement `RandomPlayer` and `KeyboardHumanPlayer`
* The returned actions from the game problem represent all posible moves and all possible series of jumps

`KeyboardHumanPlayer` vs `NonRepeatingRandomPlayer`:

```
Player to move: 0
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  2
2  .  .  .  .  .
3  1  .  .  .  .
4  1  1  .  .  .
0:	(3, 0); (2, 0)
1:	(3, 0); (2, 1)
2:	(3, 0); (3, 1)
3:	(4, 0); (3, 1)
4:	(4, 0); (2, 0)
5:	(4, 0); (4, 2)
6:	(4, 1); (3, 1)
7:	(4, 1); (3, 2)
8:	(4, 1); (4, 2)
Enter which action to take: 1
Applied move: (3, 0); (2, 1)
Player to move: 1
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  2
2  .  1  .  .  .
3  .  .  .  .  .
4  1  1  .  .  .
Applied move: (1, 4); (2, 4)
Player to move: 0
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  .
2  .  1  .  .  2
3  .  .  .  .  .
4  1  1  .  .  .
0:	(2, 1); (1, 1)
1:	(2, 1); (1, 2)
2:	(2, 1); (2, 0)
3:	(2, 1); (2, 2)
4:	(2, 1); (3, 0)
5:	(2, 1); (3, 1)
6:	(4, 0); (3, 0)
7:	(4, 0); (3, 1)
8:	(4, 0); (4, 2)
9:	(4, 1); (3, 1)
10:	(4, 1); (3, 2)
11:	(4, 1); (4, 2)
Enter which action to take: 9
Applied move: (4, 1); (3, 1)
Player to move: 1
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  .
2  .  1  .  .  2
3  .  1  .  .  .
4  1  .  .  .  .
Applied move: (2, 4); (3, 4)
Player to move: 0
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  .
2  .  1  .  .  .
3  .  1  .  .  2
4  1  .  .  .  .
0:	(2, 1); (1, 1)
1:	(2, 1); (1, 2)
2:	(2, 1); (2, 0)
3:	(2, 1); (2, 2)
4:	(2, 1); (3, 0)
5:	(2, 1); (4, 1)
6:	(3, 1); (2, 2)
7:	(3, 1); (3, 0)
8:	(3, 1); (3, 2)
9:	(3, 1); (4, 1)
10:	(3, 1); (1, 1)
11:	(4, 0); (3, 0)
12:	(4, 0); (4, 1)
13:	(4, 0); (2, 2)
14:	(4, 0); (2, 2); (2, 0)
Enter which action to take: 14
Applied move: (4, 0); (2, 2); (2, 0)
Player to move: 1
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  .
2  1  1  .  .  .
3  .  1  .  .  2
4  .  .  .  .  .
Applied move: (3, 4); (2, 4)
Player to move: 0
   0  1  2  3  4
0  .  .  .  2  2
1  .  .  .  .  .
2  1  1  .  .  2
3  .  1  .  .  .
4  .  .  .  .  .
0:	(2, 0); (1, 0)
1:	(2, 0); (1, 1)
2:	(2, 0); (3, 0)
3:	(2, 0); (2, 2)
4:	(2, 0); (2, 2); (4, 0)
5:	(2, 1); (1, 1)
6:	(2, 1); (1, 2)
7:	(2, 1); (2, 2)
8:	(2, 1); (3, 0)
9:	(2, 1); (4, 1)
10:	(3, 1); (2, 2)
11:	(3, 1); (3, 0)
12:	(3, 1); (3, 2)
13:	(3, 1); (4, 0)
14:	(3, 1); (4, 1)
15:	(3, 1); (1, 1)
Enter which action to take: 
```

### CODE REVIEW

**Comment by Zedrichu** [comment]
There is an error in the movement logic as you shouldn't move diagonally towards the goal corner

File: src/by_book/chinese_toolbox.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+       0  1  2  3  4
+    0  .  .  .  .  .
+    1  .  .  x  x  .
+    2  .  x  1  x  .
+    3  .  x  x  .  .
+    4  .  .  .  .  .
+    """
+    for i in range(-1, 2):
+        for j in range(-1, 2):
+            if i == -1 and j == -1 or i == 1 and j == 1:
```
Comment (Zedrichu) [inline src/by_book/chinese_toolbox.py]: The pairs `(-1,1)` and `(1,-1)` should be skipped instead

Comment (TermanEmil) [reply src/by_book/chinese_toolbox.py]: good catch
I've pushed a fix

Comment (Zedrichu) [reply src/by_book/chinese_toolbox.py]: Well done! Let's keep your branch as a PR until we ask the TAs how to proceed
In the meantime, I will use it to improve and combine with what we have, but still with the smaller action concept

**Pull request closed without merging**